pilot/pkg/config/kube/gateway/conversion.go

```diff
@@ -1449,8 +1450,24 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey]map[k8s.Se
 			if len(addressesToReport) == 0 {
 				// There are no external addresses, so report the internal ones
 				// TODO: should we always report both?
-				addressesToReport = internal
 				addrType = k8s.HostnameAddressType
+				for _, hostport := range internal {
+					svchost, _, _ := net.SplitHostPort(hostport)
+					svc := r.Context.ps.ServiceIndex.HostnameAndNamespace[host.Name(svchost)][obj.Namespace]
+					if svc.Attributes.Type != corev1.ServiceTypeLoadBalancer {
```

Comment:
IMO its better to have ResolveGatewayInstances incapsulate this logic in some way. Its a bit messy to have the  internal logic leak into gateway code (and a bug cause by someone refactoring - GatewayContext is suppose to encapsulate the internals of push_context and thus must live in the push context package...)